fml/raster_thread_merger_unittests.cc

```diff
@@ -616,11 +487,148 @@ TEST(RasterThreadMerger, SetMergeUnmergeCallback) {
 
   raster_thread_merger->DecrementLease();
   ASSERT_EQ(2, callbacks);
+}
 
-  term1.Signal();
-  term2.Signal();
-  thread1.join();
-  thread2.join();
+TEST(RasterThreadMerger, MultipleMergersCanMergeSameThreadPair) {
+  TaskQueueWrapper queue1;
+  TaskQueueWrapper queue2;
+  fml::TaskQueueId qid1 = queue1.GetTaskQueueId();
+  fml::TaskQueueId qid2 = queue2.GetTaskQueueId();
+  const auto raster_thread_merger1_ =
+      fml::RasterThreadMerger::CreateOrShareThreadMerger(nullptr, qid1, qid2);
+  const auto raster_thread_merger2_ =
+      fml::RasterThreadMerger::CreateOrShareThreadMerger(raster_thread_merger1_,
+                                                         qid1, qid2);
+  const int kNumFramesMerged = 5;
+  ASSERT_FALSE(raster_thread_merger1_->IsMerged());
+  ASSERT_FALSE(raster_thread_merger2_->IsMerged());
+
+  // Merge using the first merger
+  raster_thread_merger1_->MergeWithLease(kNumFramesMerged);
+
+  ASSERT_TRUE(raster_thread_merger1_->IsMerged());
+  ASSERT_TRUE(raster_thread_merger2_->IsMerged());
+
+  // let there be one more turn till the leases expire.
+  for (int i = 0; i < kNumFramesMerged - 1; i++) {
+    // Check merge state using the two merger
+    ASSERT_TRUE(raster_thread_merger1_->IsMerged());
+    ASSERT_TRUE(raster_thread_merger2_->IsMerged());
+    raster_thread_merger1_->DecrementLease();
+  }
+
+  ASSERT_TRUE(raster_thread_merger1_->IsMerged());
+  ASSERT_TRUE(raster_thread_merger2_->IsMerged());
+
+  // extend the lease once with the first merger
+  raster_thread_merger1_->ExtendLeaseTo(kNumFramesMerged);
+
+  // we will NOT last for 1 extra turn, we just set it.
+  for (int i = 0; i < kNumFramesMerged; i++) {
+    // Check merge state using the two merger
+    ASSERT_TRUE(raster_thread_merger1_->IsMerged());
+    ASSERT_TRUE(raster_thread_merger2_->IsMerged());
+    raster_thread_merger1_->DecrementLease();
+  }
```

Comment:
Is this what we wanted for the lease decrementing twice as fast issue we previously talked about?  I'm unclear on the ramifications, @iskakaushik ?

Reply:
Now decrementing won't be twice fast because they store in a `map<caller, lease_term>` in shared_merger, this case tests `merger1->Merge()`, `merger1->DecrementLease()`, `merger1->ExtendLeaseTo()` and only check merger2's `IsMerged()` is right or not.

Reply:
See below 😄